### PR TITLE
Faster Qblox integration and beyond

### PR DESCRIPTION
**Pull request by hamidelmaazouz**
- Starts to treat the instruction builder as an (linear) IR. IR in the `pass_base` module is intentionally left opaque because there will be more work **either** improving the existing instruction builder or coming up with a new IR. Will call this "QatIR". It's a different subject but slightly related to this draft. At this point we could also start looking into MLIR (and maybe its python bindings (`pyMLIR`) ?)
- Introduces pass infrastructure very much inspired from LLVM's new PassManager. Particularly, "concept-based polymorphism" is proposed to replace unnecessary inheritance hierarchies. The new infrastructure also fosters a flat architecture and plug-in paradigm where passes can be easily added and individually and thoroughly tested. For example, `XEngine.optimize()` and `XEngine.validate()` become a sequence of transform and validation passes, some Control HW limit verifications (like Lucy 90 microns problem, potential Toshiko limit checks) can be written as a pass: no need to introduce ad-hoc inheritance on the engine (see `NewQuantumRuntime` to see how dynamic dispatch is avoided). (Also see `FastQbloxEngine`, `FastQbloxEmitter`, and `CFGWalker` for codegen in the backend)
- Introduces control flow for which basic scoping for Repeat and Sweep instructions is necessary to help build the control flow graph
- New Qblox codegen as a DFS traversal on the CFG to allow for more flexibility and control. Future directions involve proper concepts around MIR and Global Instruction Selection, but this is enough for now.
- Most TODOs are technical and introduce more ideas to where this can go during the current and future iterations

### CODE REVIEW

File: src/qat/ir/pass_base.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    """
+    A base pass that mutates the IR in-place.
+    """
+
+    def run(self, ir, *args, **kwargs):
+        if args:
+            analyses = args[0]
+        else:
+            analyses = PassResultSet()
+            args = tuple([analyses] + list(args))
```
Comment (daria-oqc) [inline src/qat/ir/pass_base.py:126]:
```suggestion
            args = tuple([analyses])
```

File: src/qat/ir/pass_base.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    """
+    A base pass to validates semantics and enforce legalisation on the IR.
+    """
+
+    def run(self, ir, *args, **kwargs):
+        if args:
+            analyses = args[0]
+        else:
+            analyses = PassResultSet()
+            args = tuple([analyses] + list(args))
```
Comment (daria-oqc) [inline src/qat/ir/pass_base.py:145]:
```suggestion
            args = tuple([analyses])
```

File: src/qat/ir/pass_base.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+    def __init__(self):
+        self.passes: List[PassModel] = []
+
+    def run(self, ir, *args, **kwargs):
+        if args:
+            analyses = args[0]
+        else:
+            analyses = PassResultSet()
+            args = tuple([analyses] + list(args))
```
Comment (daria-oqc) [inline src/qat/ir/pass_base.py:176]:
```suggestion
            args = tuple([analyses])
```

File: src/qat/ir/pass_base.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+
+class AnalysisPass(PassInfoMixin):
+    """
+    A base pass to compute some form of analysis on the IR and return it.
+    The IR is imperatively left intact.
+    """
+
+    def run(self, ir, *args, **kwargs):
+        pass
```
Comment (bgsach) [inline src/qat/ir/pass_base.py:113]: What are analysis passes expected to return?

File: src/qat/ir/pass_base.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+            analyses = args[0]
+        else:
+            analyses = PassResultSet()
+            args = (analyses,)
+        self.do_run(ir, *args, **kwargs)
+        return analyses
+
+    @abstractmethod
+    def do_run(self, ir, *args, **kwargs):
+        pass
```
Comment (bgsach) [inline src/qat/ir/pass_base.py:151]: Do validation passes throw or return false when validation fails?

File: src/qat/ir/pass_base.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+            analyses = PassResultSet()
+            args = (analyses,)
+        for p in self.passes:
+            p.run(ir, *args, **kwargs)
+        return analyses
+
+    def add(self, pass_obj):
+        self.passes.append(PassModel(pass_obj))
+        return self
+
```
Comment (bgsach) [inline src/qat/ir/pass_base.py:184]:
```
    def __or__(self, pass_obj):
        return self.add(pass_obj)
```

which enables...

```
        return PassManager() |
                        PhaseOptimisation()
                        PostProcessingOptimisation()
                        CtrlHwValidation(self.engine)
                        PostProcessingValidation(self.model)
       ```

**Comment by hamidelmaazouz** [comment]
Closing this PR as it's been split into three main PRs for clarity and ease of review:
- https://github.com/oqc-community/qat/pull/216
- https://github.com/oqc-community/qat/pull/217
- https://github.com/oqc-community/qat/pull/219

**Pull request closed without merging**